### PR TITLE
feat: add metric for converter

### PR DESCRIPTION
**Pull request by Desiki-high**
**We need to collect the converter metric for nydusify, such as conversion time image size.**
Related: https://github.com/dragonflyoss/image-service/issues/1263

### CODE REVIEW

File: pkg/converter/converter.go [hunk 1]
```diff
@@ -85,14 +93,15 @@ func (cvt *Converter) pull(ctx context.Context, source string) error {
 	return nil
 }
 
-func (cvt *Converter) Convert(ctx context.Context, source, target string) error {
+func (cvt *Converter) Convert(ctx context.Context, source, target string) (Metric, error) {
```
Comment (imeoer) [inline pkg/converter/converter.go:96]: May be return `*Metric` here.

File: pkg/converter/converter.go [hunk 2]
```diff
@@ -104,21 +113,30 @@ func (cvt *Converter) Convert(ctx context.Context, source, target string) error
...
 	logger.Infof("converting image %s", source)
 	start = time.Now()
 	desc, err := cvt.driver.Convert(ctx, cvt.provider, source)
 	if err != nil {
-		return errors.Wrap(err, "convert image")
+		return metric, errors.Wrap(err, "convert image")
+	}
+	metric.ConversionElapsed = time.Since(start)
+	logger.Infof("converted image %s, elapse %s", target, metric.ConversionElapsed)
+	descs, err := images.ChildrenHandler(cvt.provider.ContentStore())(ctx, *desc)
```
Comment (imeoer) [inline pkg/converter/converter.go:133]: Can we move the size calculation to `metric.ImageSize(ctx context.Context, cs content.Store, desc ocispec.Descriptor)`.

File: pkg/converter/converter.go (hunk 1 above)
Comment (Desiki-high) [reply pkg/converter/converter.go]: fix

File: pkg/converter/converter.go (hunk 2 above)
Comment (Desiki-high) [reply pkg/converter/converter.go]: fix

File: pkg/converter/metric.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+func (metric *Metric) imageSize(ctx context.Context, cs content.Store, image *ocispec.Descriptor) (int64, error) {
+	var imageSize int64
+	switch image.MediaType {
+	case ocispec.MediaTypeImageIndex, MediaTypeDockerSchema2ManifestList:
+		index, err := images.ChildrenHandler(cs)(ctx, *image)
+		if err != nil {
+			return imageSize, err
+		}
+		for _, manifest := range index {
+			layers, err := images.ChildrenHandler(cs)(ctx, manifest)
```
Comment (imeoer) [inline pkg/converter/metric.go:70]: Maybe `layers` -> `children` ? The return includes config desc.

File: pkg/converter/metric.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	metric.SourceImageSize, err = metric.imageSize(ctx, cvt.provider.ContentStore(), image)
+	return err
+}
+
+func (metric *Metric) imageSize(ctx context.Context, cs content.Store, image *ocispec.Descriptor) (int64, error) {
+	var imageSize int64
+	switch image.MediaType {
+	case ocispec.MediaTypeImageIndex, MediaTypeDockerSchema2ManifestList:
+		index, err := images.ChildrenHandler(cs)(ctx, *image)
```
Comment (imeoer) [inline pkg/converter/metric.go:65]: `index` -> `manifests`.

File: pkg/converter/metric.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		if err != nil {
+			return imageSize, err
+		}
+		for _, manifest := range index {
+			layers, err := images.ChildrenHandler(cs)(ctx, manifest)
+			if err != nil {
+				return imageSize, err
+			}
+			for _, layer := range layers {
```
Comment (imeoer) [inline pkg/converter/metric.go:74]: `for _, desc := range descs {`

File: pkg/converter/metric.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+				imageSize += layer.Size
+			}
+		}
+	case ocispec.MediaTypeImageManifest, MediaTypeDockerSchema2Manifest:
+		layers, err := images.ChildrenHandler(cs)(ctx, *image)
```
Comment (imeoer) [inline pkg/converter/metric.go:79]: Ditto.

File: pkg/converter/metric.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		if err != nil {
+			return imageSize, err
+		}
+		for _, layer := range layers {
```
Comment (imeoer) [inline pkg/converter/metric.go:83]: Ditto.

**Review by imeoer — APPROVED** [review]

**Pull request merged**